### PR TITLE
core: thread_flags: optimize thread_flags_wait_one, fix doxygen

### PR DESCRIPTION
**Pull request by kaspar030**
No need to call expensive bitarithm_lsb() for just keeping the most significant bit in thread_flags_wait_one(). Also fix doxygen explaining wrong order.

### CODE REVIEW

**Comment by cgundogan** [comment]
ACK

File: core/thread_flags.c [hunk 1]
```diff
@@ -80,7 +80,8 @@ thread_flags_t thread_flags_wait_one(thread_flags_t mask)
     _thread_flags_wait_any(mask);
     thread_t *me = (thread_t*) sched_active_thread;
     unsigned tmp = me->flags & mask;
-    return _thread_flags_clear_atomic(me, thread_flags_clear(1 << bitarithm_lsb(tmp)));
+    tmp &= -tmp;
```
Comment (Kijewski) [inline core/thread_flags.c]: Better `~tmp + 1` so it does not break for architectures that don't use two's complement. At least in theory there could be such systems. Even with `-O0` gcc understands and optimizes the code, so there is no drawback.

Comment (cgundogan) [inline core/thread_flags.c]: hmm sounds sensible. If someone wants to compile for a PDP, then it might choose one's complement instead of 2's

**Comment by kaspar030** [comment]
- addressed @Kijewski's comment, added thread_flags_wait_one() to test application

@cgundogan ACK holds?

**Comment by kaspar030** [comment]
Just to make those PDP users happy. ;)

**Comment by cgundogan** [comment]
yes ACK upholds

**Comment by cgundogan** [comment]
do we need a second one or is the change minor enough?

File: core/thread_flags.c [hunk 2]
```diff
@@ -79,8 +79,10 @@ thread_flags_t thread_flags_wait_one(thread_flags_t mask)
 {
     _thread_flags_wait_any(mask);
     thread_t *me = (thread_t*) sched_active_thread;
-    unsigned tmp = me->flags & mask;
-    return _thread_flags_clear_atomic(me, thread_flags_clear(1 << bitarithm_lsb(tmp)));
+    thread_flags_t tmp = me->flags & mask;
+    /* clear all but least significant bit */
+    tmp &= ~tmp + 1;
```
Comment (OlegHahm) [inline core/thread_flags.c]: Can I have parens?

**Comment by OlegHahm** [comment]
I could give a second ACK if my comment is addressed.

File: core/thread_flags.c (hunk 2 above)
Comment (kaspar030) [inline core/thread_flags.c]: how would you like them?

Comment (OlegHahm) [inline core/thread_flags.c]: correctly ;-)

Comment (kaspar030) [inline core/thread_flags.c]: Yeah, how is correctly? "not" has precedence over "plus", so do I parenthise ```(~tmp) +1```?

Comment (OlegHahm) [inline core/thread_flags.c]: Well, that was my reason to ask for parentheses. If I don't how this evaluated, I cannot judge if it is correct. ;)

Comment (kaspar030) [inline core/thread_flags.c]: > Well, that was my reason to ask for parentheses. If I don't how this
> evaluated, I cannot judge if it is correct. ;)

Why not accept that the line contains magic that needs deep thinking to
understand, and thus is commented above to explain to the initiated what
it does?

Comment (OlegHahm) [inline core/thread_flags.c]: Why not creating an additional Github account and acknowledge your own PRs?

Comment (kaspar030) [inline core/thread_flags.c]: Could someone else please step in and either ACK this or explain which parens help understanding the expression?

Comment (OlegHahm) [inline core/thread_flags.c]: I guess `tmp &= (~tmp + 1);` would be okay.

**Comment by kaspar030** [comment]
- addressed comment

**Comment by OlegHahm** [comment]
ACK

**Comment by authmillenon** [comment]
Three ACKs (by two people) and a happy Murdock. Let's go.

**Pull request merged**